register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
 
 use Instruction::*;
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut l, mut rs) = state.clone();
+    while l < program.len() {
+        match program[l] {
+            Add(i, j) => {
+                *rs.entry(i).or_insert(Godel::new()) += 1;
+                l = j; 
+            },
+            Sub(i, j, k) => {
+                let v = rs.entry(i).or_insert(Godel::new()); 
+                if *v != 0 {
+                    *v -= 1;
+                    l = j;
+                } else {
+                    l = k; 
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
+fn encode_instruction(ins: &Instruction) -> Godel {
+    match ins {
+        Add(i, j) => 
+            encode_pair1(&Godel::from(2 * i), &Godel::from(*j)),
+        Sub(i, j, k) => 
+            encode_pair1(&Godel::from(2 * i + 1),
+            &encode_pair2(&Godel::from(*j), &Godel::from(*k))
+        ),
+        Halt => Godel::ZERO
+    }
 }
 // <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair1(x: &Godel, y: &Godel) -> Godel {
+    (Godel::from(2) * y + 1) << x.to_u32_wrapping()
 }
 // <x,y> = (2^x)*(2y+1)-1
-fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair2(x: &Godel, y: &Godel) -> Godel {
+    encode_pair1(x, y) - 1
 }
-fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+fn encode_list_to_godel(l: &[Godel]) -> Godel {
+    if l.is_empty() { return Godel::ZERO; } 
+    encode_pair1(&l[0], &encode_list_to_godel(&l[1..])) 
 }
-fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
-    unimplemented!();
+fn encode_program_to_list(program: &[Instruction]) -> Vec<Godel> {
+    program.iter().map(|x| encode_instruction(x)).collect()
+}
+// Returns 0 if there are no trailing zeros, else return trailing zero count (in binary)
+fn trailing_zeros_in_binary(x: &Godel) -> Godel {
+    let mut b = x.clone(); 
+    let mut c = Godel::new();
+    if b != 0 {
+        b = (b.clone() ^ (b - 1)) >> 1; 
+        while b != 0 {
+            c += 1;
+            b >>= 1;
+        }
+    }
+    c
+}
+fn decode_instruction(ins: &Godel) -> Instruction {
+    if *ins == 0 { return Halt; }
+    let (x, y) = decode_pair1(ins); 
+    let i: Godel = x.clone() / 2;
+    if x % 2 != 0 {
+        let (j, k) = decode_pair2(&y);
+        Sub(i.try_into().unwrap(), j.try_into().unwrap(), k.try_into().unwrap())
+    } else {
+        Add(i.try_into().unwrap(), y.try_into().unwrap())
+    }
 }
```

Comment:
I do find the fact that division necessarily consumes the Integer quite annoying with this crate. Might be a small improvement, but:
```
if *ins == 0 { return Halt; }
    let (x, y) = decode_pair1(ins);
    let x_odd = x.is_odd();
    let i: Godel = x / 2;
    if x_odd {
        let (j, k) = decode_pair2(&y);
        Sub(i.try_into().unwrap(), j.try_into().unwrap(), k.try_into().unwrap())
    } else {
        Add(i.try_into().unwrap(), y.try_into().unwrap())
    }
```
Prevents a copy.